### PR TITLE
HTTP Library

### LINKED ISSUE #493: Implement an HTTP Support Library
### Summary
In addition to data interchange (#513), it is very common in data science to pull data from web-based endpoints. To that end, we need to provide our users with an HTTP library that is capable of communicating with such endpoints.

### Value
Users will be able work with HTTP endpoints. 

### Specification

- [ ] Implement the functionality for issuing HTTP requests across the generic set.
- [ ] This should initially only support HTTP/1.1.

### Acceptance Criteria & Test Cases

- The above specification has been implemented.
- The above functionality has been rigorously tested.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #493 

Changelog:
- add: `Base.Net.Http` library
- configure CI tests

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: distribution/std-lib/Base/src/Net/Http.enso [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    connect : Request.To_Request -> Response
+
+    ## Create a request
+    request : Request.To_Request -> Response
+
+## Open an Http connection using `Uri`.
+
+   > Example
+     Open a connection and send a GET request.
+         Http.new "http://httpbin.org" http->
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Net/Http.enso]: I don't understand this pattern. What does `new` do that it requires `http` to be wrapped in a lambda? Surely it doesn't open the connection yet, so there's no underlying resource, so why not just return the builder?

File: distribution/std-lib/Base/src/Net/Http.enso [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+     Open a connection and send a Post request.
+         Http.new "http://httpbin.org" http->
+             req = Request.new Method.POST "http://httpbin.org/post" . with_body [Request.Part "key" "value"] . with_header (Header.header "X-Trace-Id" "123456789")
+             res = http.request req
+             res.code
+
+   > Example
+     Open a connection and send multiple requests.
+         Http.new "http://httpbin.org" http->
+             http.get "http://httpbin.org/get"
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Net/Http.enso]: Again, what's the point of this `http->`?

File: distribution/std-lib/Base/src/Net/Http/Request.enso [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    to_post_data : Any
+    to_post_data = this.internal_to_post_data
+
+## Create a GET request from Uri string
+
+   > Example
+     Create a GET request.
+         "http://httpbin.org/get".to_request
+Text.to_request : Request
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Net/Http/Request.enso]: this looks unnecessary, passing the string as an argument is much cleaner and equally (or more) readable

File: distribution/std-lib/Base/src/Net/Http.enso (hunk 1 above)
Comment (4e6) [reply distribution/std-lib/Base/src/Net/Http.enso]: It was inspired by Ruby, and I got the idea that it'll manage the connection. You're right, the lambda is not necessary here :+1:

File: distribution/std-lib/Base/src/Net/Http.enso [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+type Http
+
+    type Http timeout auth follow_redirects proxy
+
+    ## Send an OPTIONS request.
+
+       > Example
+         Send an OPTIONS request.
+             Http.new.options "http://httpbin.org"
```
Comment (radeusgd) [inline distribution/std-lib/Base/src/Net/Http.enso:33]: Is there a particular reason that examples point to `http` instead of `https`?

Comment (4e6) [reply distribution/std-lib/Base/src/Net/Http.enso:33]: No. I didn't think about that

**Review by kustosz — APPROVED** [review]
This is a very nice API already, a few nitpicks inline.
Additionaly, I'd like to discuss 2 improvements – LMK what you think.
1. Having module-level methods such as `Http.get uri="http://...."` would be nice. For most (layman) uses, there's little utility in the actual Http object, so this would be a nifty shortcut.
2. I kinda dislike the way this handles authorization. Basic auth is just a header, right? And also, from a programmer perspective, auth is pretty much write-only (i.e. it's unlikely you want to go request.get_auth.do_something). Therefore I'd get rid of the "specialness" of Auth and just implemented it as header helpers on request (removing the auth datatype altogether). Given that we won't predict _all_ modes of authentication, this mode of operation is much more extensive. A user can just define their own `Request.with_auth_bearer` or `Request.with_auth_oauth` or whatever else they need when connecting to a 20y old SOAP service :D.

**Comment by 4e6** [comment]
Agree with all the comments.

Changelog:
- add: helper request methods in `Http` module
- add: basic auth header
- remove: `Http.Auth` module

**Pull request merged**